csharp/ql/src/semmle/code/csharp/Conversion.qll

```diff
@@ -78,125 +92,213 @@ private Type getTypeArgument(UnboundGenericType ugt, ConstructedType ct, int i,
-  exists(Type toTypeArgument |
-    convIdentityStrictConstructedTypeFromZeroAux(ugt, fromType, i, toTypeArgument) and
-    toTypeArgument = getTypeArgument(ugt, toType, i, _) and
-    fromType != toType |
-    i = 0
-    or
-    convIdentityStrictConstructedTypeFromZero(ugt, fromType, toType, i - 1)
-  )
-}
+  pragma[nomagic]
+  private predicate convTypeArgumentsSomeUnbound(UnboundGenericType ugt, TypeArgument fromTypeArgument, TypeArgument toTypeArgument, int i) {
+    convTypeArguments(fromTypeArgument, toTypeArgument, i) and
+    fromTypeArgument = getTypeArgumentRanked(ugt, _, i)
+  }
 
-pragma [nomagic]
-private predicate convIdentityStrictConstructedTypeFromZeroAux(UnboundGenericType ugt, IdentityConvertibleConstructedType fromType, int i, Type toTypeArgument) {
-  exists(Type fromTypeArgument |
-    fromTypeArgument = getTypeArgument(ugt, fromType, i, _) and
-    convIdentityTypeArgument(fromTypeArgument, toTypeArgument)
-  )
-}
+  /**
+   * Holds if `fromTypeArgument` is identity convertible to `toTypeArgument` and
+   * both types are the `i`th type argument in _some_ constructed type
+   * based on unbound generic type `ugt`.
+   */
+  pragma[noinline]
+  private predicate convTypeArgumentsSameUnbound(UnboundGenericType ugt, TypeArgument fromTypeArgument, TypeArgument toTypeArgument, int i) {
+    convTypeArgumentsSomeUnbound(ugt, fromTypeArgument, toTypeArgument, i) and
+    toTypeArgument = getTypeArgumentRanked(ugt, _, i)
+  }
 
-private predicate convIdentityTypeArgument(TypeArgument fromType, TypeArgument toType) {
-  convIdentity(fromType, toType)
-}
+  pragma[nomagic]
+  private predicate convIdentitySingle0(UnboundGenericType ugt, IdentityConvertibleConstructedType toType, TypeArgument fromTypeArgument, TypeArgument toTypeArgument) {
+    convTypeArgumentsSameUnbound(ugt, fromTypeArgument, toTypeArgument, 0) and
+    toTypeArgument = getTypeArgumentRanked(ugt, toType, 0) and
+    ugt.getNumberOfTypeParameters() = 1
+  }
 
-/** A type that is an argument in a constructed type. */
-private class TypeArgument extends Type {
-  TypeArgument() {
-    this = any(ConstructedType ct).getATypeArgument()
+  /**
+   * Holds if the type arguments of types `fromType` and `toType` are identity
+   * convertible, and the number of type arguments is 1.
+   */
+  predicate convIdentitySingle(UnboundGenericType ugt, IdentityConvertibleConstructedType fromType, IdentityConvertibleConstructedType toType) {
+    exists(TypeArgument fromTypeArgument, TypeArgument toTypeArgument |
+      convIdentitySingle0(ugt, toType, fromTypeArgument, toTypeArgument) |
+      fromTypeArgument = getTypeArgumentRanked(ugt, fromType, 0)
+    )
+  }
+
+  pragma[nomagic]
+  private predicate convIdentityMultiple01Aux0(UnboundGenericType ugt, IdentityConvertibleConstructedType toType, TypeArgument fromTypeArgument0, TypeArgument toTypeArgument0, TypeArgument toTypeArgument1) {
```

Comment:
Here, and elsewhere - max line width exceeded - sorry! But, would probably read more clearly.

Reply:
Is it ok to let the QL autoformatter handle this? I believe we should have that quite soon.